rabbitmq/tests/test_openmetrics.py

```diff
@@ -1,24 +1,235 @@
+            dict(name='rabbitmq.queue.messages.ram', value=1, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.queue.messages.ram_bytes', value=12, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.queue.messages.ready', value=1, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.queue.messages.ready_bytes', value=12, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.queue.messages.ready_ram', value=1, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.queue.messages.unacked', value=0, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.queue.messages.unacked_bytes', value=0, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.queue.messages.unacked_ram', value=0, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.queue.process.memory_bytes', value=104736, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.queues', value=3, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.raft.entry_commit_latency_seconds', value=0, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.raft.log_commit_index', value=0, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.raft.log_last_applied_index', value=0, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.raft.log_last_written_index', value=0, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.raft.log_snapshot_index', value=0, metric_type=aggregator.GAUGE),
+            dict(name='rabbitmq.resident_memory_limit_bytes', value=3293159424, metric_type=aggregator.GAUGE),
+        ]
+        + [
+            dict(
+                name='rabbitmq.erlang.mnesia.tablewise_size',
+                value=0,
+                metric_type=aggregator.GAUGE,
+                tags=[f"table:{table}"],
+            )
+            for table in [
+                'gm_group',
+                'rabbit_route',
+                'rabbit_index_route',
+                'rabbit_reverse_route',
+                'rabbit_durable_route',
+                'rabbit_durable_queue',
+                'rabbit_topic_trie_node',
+                'rabbit_topic_trie_edge',
+                'rabbit_exchange_serial',
+                'mirrored_sup_childspec',
+                'rabbit_topic_permission',
+                'rabbit_topic_trie_binding',
+                'rabbit_semi_durable_route',
+            ]
+        ]
+        + [
+            dict(
+                name='rabbitmq.erlang.vm.memory.system_bytes',
+                value=v,
+                metric_type=aggregator.GAUGE,
+                tags=[f"usage:{usage}"],
+            )
+            for usage, v in [
+                ("atom", 1466705),
+                ("binary", 2896904),
+                ("code", 32222879),
+                ("ets", 3845072),
+                ("other", 16062681),
+            ]
+        ]
+        # TODO: This doesn't match even though all the fields are the same.
```

Comment:
Here's the error I get:

```shell
  E   AssertionError: Needed at least 1 candidates for 'rabbitmq.erlang.vm.allocators', got 0
E   Expected:
E           MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_node:1', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   Similar submitted:
E   Score   Most similar
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:1', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:5', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:4', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:3', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:2', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:0', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:ll_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:1', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:std_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:1', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:ets_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:1', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:5', 'kind:mbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:4', 'kind:mbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:2', 'kind:mbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:ll_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:5', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:ll_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:4', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:ll_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:3', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
```

Reply:
the value is a float `0.0`, not an integer `0`

Reply:
also : `instance_node` -> `instance_no` it seems

Reply:
Ah, the tag is a good catch, thanks 🙈 

I haven't had problems matching floats and ints in other metrics, so I'm not sure that's an issue.

In any case, it's still not matching:

```shell
E   AssertionError: Needed at least 1 candidates for 'rabbitmq.erlang.vm.allocators', got 0
E   Expected:
E           MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:1', 'kind:mbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   Similar submitted:
E   Score   Most similar
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:5', 'kind:mbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:4', 'kind:mbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:2', 'kind:mbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:1', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:5', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:4', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:3', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:2', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:0', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:ll_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:5', 'kind:mbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:ll_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:1', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:sl_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:1', 'kind:mbcs_pool', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:std_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:1', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:ets_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:5', 'kind:mbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
E   1.00    MetricStub(name='rabbitmq.erlang.vm.allocators', type=0, value=0.0, tags=['alloc:ets_alloc', 'endpoint:localhost:15692/metrics', 'instance_no:1', 'kind:sbcs', 'usage:blocks'], hostname='', device=None, flush_first_value=False)
```

Reply:
There's no metrics with both 'instance_no:5' and 'kind:mbcs' I think

Reply:
Oh, i see the problem! I'll reopen this discussion if I don't resolve it, but I'm pretty sure I know what it is.